### ISSUE TITLE
Deadlock when result in select is modified by another running transaction

### ISSUE THREAD

**Issue by vangheem** [cockroachdb/cockroach#16091]
I could be misunderstanding something here about how transactions work differently in cockroach to postgresql but I've been hitting my head against this issue for some time now.

**BUG REPORT**

This is using basic setup with single node and running with docker on version 1.0 of cockroachdb.

Steps to reproduce:
 1. start transaction
 2. add row in transaction
 3. in another transaction-less connection, query table that a row was added to in step 2
 4. commit transaction

Expected result:

- step 3 to return data with the current state of the database before commit is done with other transaction

What happens:

- step 3 hangs because, it seems to me, cockroach thinks it needs to wait for the transaction to be committed before it'll finish

Example python script using asyncpg to demonstrate the issue: 
```
import asyncpg
import asyncio


async def run():
    # Establish a connection to an existing database named "test"
    # as a "postgres" user.
    pool = await asyncpg.create_pool(
        dsn='postgresql://root@localhost:26257?sslmode=disable',
        max_size=5,
        min_size=2)
    conn1 = await pool.acquire()
    conn2 = await pool.acquire()

    # create db/tables...
    await conn2.execute('create database if not exists testdb;')
    await conn2.execute('set database to testdb;')
    await conn2.execute('create table if not exists foobar (foo INT, bar TEXT);')
    await conn1.execute('set database to testdb;')

    txn = conn1.transaction()
    await txn.start()

    _id = await conn2.fetchval('''SELECT unique_rowid()''')

    await conn1.execute('''INSERT INTO foobar (foo, bar)
                           VALUES ($1::int, $2::text)''', _id, 'foobar')

    # Execute a statement to create a new table.
    await conn2.fetch('''SELECT MAX(foo) from foobar;''')
    await conn2.close()

    await txn.rollback()
    await conn1.close()


if __name__ == '__main__':
    event_loop = asyncio.get_event_loop()
    event_loop.run_until_complete(run())
```

**Comment by vangheem** [comment]
For anyone that comes across this, if you RTM, you'll notice the documentation on transactions touches on this:

https://www.cockroachlabs.com/docs/transactions.html

> When two transactions contend for the same resources indirectly, they may create a dependency cycle leading to a deadlock situation, where both transactions are waiting on the other to finish. In these cases, CockroachDB allows the transaction with higher priority to abort the other, which must then retry. On retry, the transaction inherits the higher priority. This means that each retry makes a transaction more likely to succeed in the event it again experiences deadlock.

**Comment by bdarnell** [comment]
This isn't a deadlock, it's just a lock. `conn1` has a lock on the row it is writing to (effectively - our transaction locking model is a hybrid of optimistic and pessimistic concurrency and it's not exactly a lock, but it's close enough to one for discussion purposes). `conn1` could proceed freely if it were being run in its own task. In a deadlock, we'd have locks in both directions: `conn1` cannot proceed until `conn2` releases its locks, and `conn2` cannot proceed until `conn1` releases its locks. In that case, the server would have to abort one of them.

I think the difference between CockroachDB and PostgreSQL here has to do with the transaction isolation levels: CockroachDB defaults to `SERIALIZABLE` while PostgreSQL defaults to `READ COMMITTED`. If you changed the transaction isolation level to `SERIALIZABLE` on postgres, I think you'd see the same blocking behavior there.

**Comment by vangheem** [comment]
Yes, I understand it better now. Thank you for the response.
